Add Stair Wrench

hugeblank opened the pull request:
![image](https://github.com/SwitchCraftCC/sc-goodies/assets/16393543/25e6241a-15bb-4ce6-921d-efd4f8a4a002)

Adds the stair wrench. Grants the ability to modify the state of a stair to otherwise impossible (but not unreasonable!) orientations. Has 3 modes that correspond with the block properties of the stair: facing, half, and shape. Mode is preserved in the tool. Can be used by turtles. Has not been tested with claimkit. 

Programmer art texture. Baby's first experience with Kotlin code.

hugeblank committed:
add stair wrench

emmachase commented on src/main/kotlin/io/sc3/goodies/misc/StairWrenchItem.kt:
```diff
@@ -0,0 +1,94 @@
+import net.minecraft.world.WorldAccess
+
+class StairWrenchItem(settings: Settings) : BaseItem(settings) {
+  private var properties = listOf(StairsBlock.FACING, StairsBlock.HALF, StairsBlock.SHAPE) as List<Property<*>>
```
Why are we casting here? Just set the type on the declaration, or specify the listOf type parameter.

emmachase commented on src/main/kotlin/io/sc3/goodies/misc/StairWrenchItem.kt:
```diff
@@ -0,0 +1,94 @@
+
+  override fun canMine(state: BlockState?, world: World, pos: BlockPos?, miner: PlayerEntity): Boolean {
+    if (!world.isClient) {
+      this.use(miner, state!!, world, pos!!, false, miner.getStackInHand(Hand.MAIN_HAND))
```
We should avoid non-null assertions. Although unlikely, presumably they are nullable for a reason.

emmachase replied:
```suggestion
      this.use(miner, state ?: return false, world, pos ?: return false, false, miner.getStackInHand(Hand.MAIN_HAND))
```

emmachase commented on src/main/kotlin/io/sc3/goodies/misc/StairWrenchItem.kt:
```diff
@@ -0,0 +1,94 @@
+
+
+  private fun sendMessage(player: PlayerEntity, message: Text) {
+    (player as ServerPlayerEntity).sendMessageToClient(message, true)
```
Why not make sendMessage take ServerPlayerEntity?

emmachase commented on src/main/kotlin/io/sc3/goodies/misc/StairWrenchItem.kt:
```diff
@@ -0,0 +1,94 @@
+    if (block is StairsBlock) {
+      val stateManager = block.stateManager
+      val mode = stack.orCreateNbt.getString("Property")
+      var property = stateManager.getProperty(mode)
```
This being a var seems unnecessary to me. Just make new variables if necessary, why mutate?

emmachase commented on src/main/kotlin/io/sc3/goodies/misc/StairWrenchItem.kt:
```diff
@@ -0,0 +1,94 @@
+          property = properties.iterator().next()
+        }
+        val blockState = cycle(state, property, player.shouldCancelInteraction())
+        world.setBlockState(pos, blockState, 18)
```
What's 18?

hugeblank replied:
it's what the debug stick uses and I couldn't be bothered to figure it out. Mainly because I'm not sure where the proper flag fields are hiding

hugeblank replied:
found them, they were in `Block`, 18 is `FORCE_STATE | NOTIFY_LISTENERS`

emmachase commented on src/main/kotlin/io/sc3/goodies/misc/StairWrenchItem.kt:
```diff
@@ -0,0 +1,94 @@
+        val blockState = cycle(state, property, player.shouldCancelInteraction())
+        world.setBlockState(pos, blockState, 18)
+      } else {
+        property = cycle(properties, property, player.shouldCancelInteraction()) as Property<*>
```
This cast shouldn't be necessary, as it doesn't tighten the upper bound of the type at all?

emmachase replied:
Is it possible for property to be null like in the other branch? Your signature for `cycle` doesn't seem to accept a nullable type.

emmachase commented on src/main/kotlin/io/sc3/goodies/misc/StairWrenchItem.kt:
```diff
@@ -0,0 +1,94 @@
+      var property = stateManager.getProperty(mode)
+      if (update) {
+        if (property == null) {
+          property = properties.iterator().next()
```
```suggestion
          property = properties.first()
```

emmachase replied:
Also shouldn't this just be hoisted to the declaration?
```
val property = stateManager.getProperty(mode) ?: properties.first()
```

hugeblank committed:
implement requested changes

Lemmmy commented on src/main/kotlin/io/sc3/goodies/datagen/recipes/handlers/StairWrenchRecipe.kt:
```diff
@@ -0,0 +1,27 @@
+import net.minecraft.registry.tag.ItemTags.STAIRS
+import java.util.function.Consumer
+
+object StairWrenchRecipe: RecipeHandler { // Having a single file for a recipe feels like a sin
```
deal with it

hugeblank replied:
uwaaa

Lemmmy commented on src/main/kotlin/io/sc3/goodies/misc/StairWrenchItem.kt:
```diff
@@ -0,0 +1,88 @@
+import net.minecraft.world.WorldAccess
+
+class StairWrenchItem(settings: Settings) : BaseItem(settings) {
+  private var properties: List<Property<*>> = listOf(StairsBlock.FACING, StairsBlock.HALF, StairsBlock.SHAPE)
```
back to this one, is it possible to remove the type entirely? either:
```suggestion
  private var properties = listOf(StairsBlock.FACING, StairsBlock.HALF, StairsBlock.SHAPE)
```

or if not:

```suggestion
  private var properties = listOf<Property<*>>(StairsBlock.FACING, StairsBlock.HALF, StairsBlock.SHAPE)
```

just trying to think of making this line less verbose

hugeblank replied:
the linter doesn't yell at me so presumably not, will test.

Lemmmy commented on src/main/kotlin/io/sc3/goodies/misc/StairWrenchItem.kt:
```diff
@@ -0,0 +1,88 @@
+class StairWrenchItem(settings: Settings) : BaseItem(settings) {
+  private var properties: List<Property<*>> = listOf(StairsBlock.FACING, StairsBlock.HALF, StairsBlock.SHAPE)
+
+  override fun canMine(state: BlockState?, world: World, pos: BlockPos?, miner: PlayerEntity): Boolean {
+    if (!world.isClient) {
+      this.use(miner, state ?: return false, world, pos ?: return false, false, miner.getStackInHand(Hand.MAIN_HAND))
```
a quick look at the two places `canMine` is called shows that both `state` and `pos` are definitely non-null

```suggestion
  override fun canMine(state: BlockState, world: World, pos: BlockPos, miner: PlayerEntity): Boolean {
    if (!world.isClient) {
      this.use(miner, state, world, pos, false, miner.getStackInHand(Hand.MAIN_HAND))
```

hugeblank replied:
learning the lesson that the java to kotlin converter has no idea what it's doing (which is bad because neither do I)

Lemmmy replied:
it's not that it doesn't know what it's doing, it's just that it's assuming that types are nullable when they aren't explicitly marked as nullable or non-null, and it can't be inferred. it always errs on the safe side, because otherwise it'll lead to runtime NPEs

Lemmmy commented on src/main/kotlin/io/sc3/goodies/misc/StairWrenchItem.kt:
```diff
@@ -0,0 +1,88 @@
+  }
+
+  private fun use(player: PlayerEntity, state: BlockState, world: WorldAccess, pos: BlockPos, update: Boolean, stack: ItemStack): Boolean {
+    val block = state.block
+    if (block is StairsBlock) {
```
not sure if relying on the `StairsBlock` class is the right move here (as opposed to, say, a tag), but if it is, consider removing the extra level of indentation by doing this instead:

```suggestion
    val block = state.block as? StairsBlock ?: return false
```

Lemmmy replied:
the other argument for a tag is that i can see people asking for this to support other types of blocks in the future, which this would make easier to do. though idk how it'd play with the different modes you've specified when said blocks have different states

hugeblank replied:
idea suggested 
```kt
if (state.block !is StairsBlock) return false
```
but I'll look into the tag way. My only concern is that blocks tagged as stairs may not necessarily have the 3 properties. Checking the class is the safest way.

Lemmmy commented on src/main/kotlin/io/sc3/goodies/misc/StairWrenchItem.kt:
```diff
@@ -0,0 +1,88 @@
+    return ActionResult.success(world.isClient)
+  }
+
+  private fun use(player: PlayerEntity, state: BlockState, world: WorldAccess, pos: BlockPos, update: Boolean, stack: ItemStack): Boolean {
```
i'm concerned that this method name confusingly shadows `Item.use`, despite not being an overload for it - is there another name that would be reasonable?

hugeblank committed:
operate on stair tags & force facing cycle clockwise

Lemmmy commented on src/main/kotlin/io/sc3/goodies/misc/StairWrenchItem.kt:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  private fun crank(player: PlayerEntity, state: BlockState, world: WorldAccess, pos: BlockPos, update: Boolean, stack: ItemStack): Boolean {
+    if (!state.streamTags().anyMatch {
+      it.id.equals(Identifier("minecraft:stairs"))
+    }) return false
```
huh??
```suggestion
    if (!state.isIn(BlockTags.STAIRS)) return false
```

hugeblank replied:
LMAO, oh. well it was good to learn how lambdas work! I totally forgot about isIn.

Lemmmy commented on src/main/kotlin/io/sc3/goodies/misc/StairWrenchItem.kt:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  private fun <T>cycle(elements: Iterable<T>, current: T, inverse: Boolean): T {
+    @Suppress("UNCHECKED_CAST") // Suppress deez unchecked nuts, etc etc
```
remove this comment

hugeblank committed:
replace really bad tag check with the right one

Lemmmy committed:
chore: formatting + comment pass

Lemmmy committed:
fix: slight change to stair wrench tooltip

Lemmmy committed:
fix: don't mutate nbt in appendTooltip

Lemmmy committed:
fix: warn player about invalid wrench properties

This won't have any effect right now, but a bit of future-proofing UX.

Lemmmy committed:
fix: make it work in singleplayer

Replaces left-click behaviour (debug stick) with right-click and
shift-right-click. Removes the inverse property cycling for this, which
isn't a huge loss since there's only three
properties anyway.

Lemmmy committed:
feat: add custom stairs/slabs to vanilla tags
